File: src/main/scala/de/mineformers/drash/api/structure/Structure.java
package de.mineformers.drash.api.structure;

import com.google.common.collect.ImmutableList;
import net.minecraft.block.Block;
import net.minecraft.nbt.NBTTagCompound;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * Structure
 *
 * @author PaleoCrafter
 */
public class Structure
{
    private List<Layer> layers;
    private List<NBTTagCompound> entities;

    public Structure()
    {
        layers = new LinkedList<>();
        entities = new ArrayList<>();
    }

    public void addLayer(Layer layer)
    {
        if (layer == null) throw new IllegalArgumentException("Layer may not be null");
        layers.add(layer);
    }

    public Layer.BlockInfo getBlock(int x, int y, int z)
    {
        Layer layer = getLayer(y);
        if(layer == null)
            return null;
        return layer.get(x, z);
    }

    public void setBlock(int x, int y, int z, Block block, int metadata, NBTTagCompound tile)
    {
        getLayer(y).set(x, z, block, metadata, tile);
    }

    public Layer getLayer(int y)
    {
        if (y >= getHeight() || y < 0)
            return null;
        return layers.get(y);
    }

    public void addEntity(NBTTagCompound tag)
    {
        entities.add(tag);
    }

    public void clear()
    {
        layers.clear();
    }

    public int getWidth()
    {
        int widest = 0;
        for (Layer layer : layers)
            if (widest < layer.getWidth())
                widest = layer.getWidth();
        return widest;
    }

    public int getLength()
    {
        int longest = 0;
        for (Layer layer : layers)
            if (longest < layer.getLength())
                longest = layer.getLength();
        return longest;
    }

    public int getHeight()
    {
        return layers.size();
    }

    public final List<Layer> getLayers()
    {
        return ImmutableList.copyOf(layers);
    }

    public final List<NBTTagCompound> getEntities()
    {
        return ImmutableList.copyOf(entities);
    }

    public Structure copy()
    {
        Structure struct = new Structure();
        for (Layer layer : layers)
            struct.addLayer(layer.copy());
        return struct;
    }

    @Override
    public String toString()
    {
        return "Structure=" + layers.toString();
    }
}

File: src/main/scala/de/mineformers/drash/api/structure/Structure.scala
package de.mineformers.drash.api.structure

import net.minecraft.nbt.NBTTagCompound
import net.minecraft.block.Block
import scala.collection.mutable
import de.mineformers.drash.structure.StructureWorld
import net.minecraft.util.AxisAlignedBB

/**
 * Structure
 *
 * @author PaleoCrafter
 */
class Structure {
  private val layers = mutable.ListBuffer.empty[Layer]
  private val entities = mutable.ListBuffer.empty[NBTTagCompound]

  def addLayer(layer: Layer): Unit = {
    if (layer == null) throw new IllegalArgumentException("Layer may not be null")
    layers append layer
  }

  def getBlock(x: Int, y: Int, z: Int): BlockInfo = {
    val layer: Layer = getLayer(y)
    if (layer == null) return null
    layer.get(x, z)
  }

  def setBlock(x: Int, y: Int, z: Int, block: Block, metadata: Int, tile: NBTTagCompound): Unit = {
    getLayer(y).set(x, z, block, metadata, tile)
  }

  def getLayer(y: Int): Layer = {
    if (y >= getHeight || y < 0) return null
    layers(y)
  }

  def addEntity(tag: NBTTagCompound): Unit = {
    entities append tag
  }

  def clear(): Unit = {
    layers.clear()
  }

  def getWidth: Int = {
    var widest: Int = 0
    for (layer <- layers) if (widest < layer.width) widest = layer.width
    widest
  }

  def getLength: Int = {
    var longest: Int = 0
    for (layer <- layers) if (longest < layer.length) longest = layer.length
    longest
  }

  def getHeight: Int = layers.size

  def update(world: StructureWorld): Unit = {
    for(y <- 0 until getHeight; x <- 0 until getWidth; z <- 0 until getLength) {
      val info = getBlock(x, y, z)
      if(info != null)
        info.update(world, y)
    }
  }

  final def getLayers: List[Layer] = layers.toList

  final def getEntities: List[NBTTagCompound] = entities.toList

  def copy: Structure = {
    val struct: Structure = new Structure
    for (layer <- layers) struct.addLayer(layer.copy)
    struct
  }

  override def toString: String = "Structure=" + layers.toString
}